File: bubble_sort.go
package clrs

func bubbleSort(a []int) {
	for i := 0; i <= len(a)-2; i++ {
		for j := len(a) - 1; j >= i+1; j-- {
			if a[j] < a[j-1] {
				a[j], a[j-1] = a[j-1], a[j]
			}
		}
	}
}

File: src/ch02/bubblesort.rs
use std::cmp;
use super::Value;

pub fn bubblesort<T>(a: &mut [Value<T>])
    where T: cmp::PartialOrd + Copy
{
    let n = a.len();
    if n < 2 {
        return;
    }
    for i in 0..n - 1 {
        for j in (i + 1..n).rev() {
            if a[j] < a[j - 1] {
                a.swap(j, j - 1);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ::test_cases;

    #[test]
    fn test_bubblesort() {
        macro_rules! test {
            ( $( $i:tt ),* ) => {{
                $(
                    let (mut a, want) = test_cases::SORT.$i;
                    bubblesort(&mut a);
                    assert_eq!(want, a);
                )*
            }};
        }
        test!(0, 1, 2, 3, 4, 5, 6, 7);
    }
}
